source/CIS/cis-4.rst

```diff
@@ -0,0 +1,733 @@
+^^^^^^^^^^^^^^^^^^^^^^^^^^
+
+Revoke a credential by the issuer's request.
+The issuer is authorized to revoke a credential if the transaction sender's address is the same as the return value of :ref:`CIS-4-functions-issuer`.
+
+Parameter
+~~~~~~~~~
+
+The parameter is the credential ID :ref:`CIS-4-CredentialHolderId` and optional string indicating the revocation reason.
+
+It is serialized as :ref:`CIS-4-CredentialHolderId` followed by 1 byte to indicate whether a reason is included.
+If its value is 0, then no reason string is present, if the value is 1 then the bytes corresponding to the reason string follow::
+
+  OptionalReason ::= (0: Byte)
+                   | (1: Byte) (n: Byte) (reason_string: Byteⁿ)
+  RevokeCredentialIssuerParam ::= (credential_id: CredentialHolderId) (reason: OptionReason)
+
+.. TODO: what kind of characters are allowed? ASCII, Unicode?
+
+Requirements
+~~~~~~~~~~~~
+
+- If revoked successfully, the credential status MUST change to ``Revoked`` (see :ref:`CIS-4-functions-credentialStatus`).
+- The revocation MUST fail if:
+    - The sender of the transaction is not the issuer.
+    - The credential ID is not present in the registry.
+    - The credential status is not one of ``Active`` or ``NotActivated`` (see :ref:`CIS-4-functions-credentialStatus`).
+
+.. _CIS-4-functions-revokeCredentialHolder:
+
+``revokeCredentialHolder``
+^^^^^^^^^^^^^^^^^^^^^^^^^^
+
+Revoke a credential by the holders's request.
+
+The holder is authorized to revoke a credential by verifying the signature with the holder's public key.
+It replaces the authorization based on checking the transaction sender address with signature verification.
+The public key is part of :ref:`CIS-4-CredentialInfo` that is used when registering a credential with the :ref:`CIS-4-functions-registerCredential` entrypoint.
+
+Parameter
+~~~~~~~~~
+
+It is serialized as :ref:`CIS-4-SignatureEd25519` (``signature``) and message data ``RevocationDataHolder`` consisting of :ref:`CIS-4-CredentialHolderId` (``credential_id``), metadata about the signature :ref:`CIS-4-SigningData` (``signing_data``), and an optional revocation reason (``reason``), serialized similarly to :ref:`CIS-4-functions-revokeCredentialIssuer`::
+
+  RevocationDataHolder ::= (credential_id: CredentialHolderId) (signing_data: SigningData) (reason: OptionReason)
+  RevokeCredentialHolderParam ::= (signature: SignatureEd25519) (data : RevocationDataHolder)
+
+
+Requirements
+~~~~~~~~~~~~
+
+- If revoked successfully, the credential status MUST change to ``Revoked`` (see :ref:`CIS-4-functions-credentialStatus`).
+- The ``RevokeCredentialHolderParam``'s ``signing_data`` MUST include a nonce to protect against replay attacks.
+  The holders's nonce is sequentially increased every time a revocation request is successfully executed.
+  The function MUST only accept a ``RevokeCredentialHolderParam`` if it has the next nonce following the sequential order.
+- The revocation MUST fail if:
```

Comment:
Make it clear that the list is an "or"., i.e., `must fail if any of the following conditions are met`.

Reply:
Done.